### ISSUE TITLE
Sometimes to overzealous in regards to unquoted variables

### ISSUE THREAD

**Issue by FransUrbo** [koalaman/shellcheck#380]
I have the code:

```
        # Unset all FSTAB_* variables
        unset $(env | grep ^FSTAB_ | sed 's,=.*,,')
```

Here, I _want_ word splitting, but <code>shellcheck</code> is complaining (giving me a warning) about this.

Also the following code is on purpose:

```
#!/bin/sh                                                                                                                                                     

my_function() {
    local CMD="$1"; shift
    local ARGS="$*"

    "$CMD" $ARGS
}

ARGS="-C -F -l -a"
my_function /bin/ls "$ARGS"
```

Gives <code>/tmp/tst.sh:7:12: note: Double quote to prevent globbing and word splitting. [SC2086]</code>.

However, if I put double quotes around <code>$ARGS</code> in the function (to silence <code>shellcheck</code>:

```
# /tmp/tst.sh 
/bin/ls: invalid option -- ' '
Try '/bin/ls --help' for more information.
```

This time only <code>note</code>, but it's still wrong. This because it interprets the <code>-C -F -l -a</code> part as ONE command, not FOUR.

**Comment by deryni** [comment]
Putting command arguments in a string is an anti-pattern. It fails to work in many cases. You don't want to do it. Use an array. (See http://mywiki.wooledge.org/BashFAQ/050 for discussion of this topic.)

A simpler (and safer) way to do what you want in the first snippet (assuming bash and that you want local as well as exported variables unset) would be to use `unset -v "${!FSTAB_*}"`. Which uses a [bash parameter expansion](http://www.gnu.org/software/bash/manual/bashref.html#Shell-Parameter-Expansion) for "Expands to the names of variables whose names begin with prefix".

**Comment by brother** [comment]
Hi Turbo! =)
When you have reviewed the warnings and debated if you want to change some of them to other solutions that might be better for one reason or other you can slap a directive to hide the error in the future.
See https://github.com/koalaman/shellcheck/wiki/Directive.

**Comment by FransUrbo** [comment]
Yes, I saw that. But that, on the other hand, is suppressing ALL message about this in the whole file. And many of them are valid. Just not those two examples I've mentioned.

**Comment by FransUrbo** [comment]
… or in the whole function (didn't notice the first comment part - nice touch!).

**Comment by FransUrbo** [comment]
Unfortunately, I honestly don't know what I expect that you do about this… In the function example, the call to <code>$CMD</code> SHOULD be quoted, but not the args.

But if we consider something like this:

```
#!/bin/sh                                                                                                                                                     

my_function() {
    local PIDFILE="$1"  ; shift
    local CMD="$1"      ; shift
    local ARGS="$*"

    "$CMD" -p "$PIDFILE" $ARGS
}

ARGS="-C -F -l -a"
my_function /bin/ls "$ARGS"
```

(ok, so this doesn't make much sense, since I'm calling it with <code>ls</code> but still, I'm sure you get the intention). Here, both <code>$CMD</code> and <code>$PIDFILE</code> must/should be escaped, but NOT <code>ARGS</code>. I have no idea how to programmatically distinguish between the two...

**Comment by FransUrbo** [comment]
POSSIBLY (just spit-balling here), the comment directive in your example could be extended to ignore a variable?

Like:

``
my_function() {
    local PIDFILE="$1"  ; shift
    local CMD="$1"      ; shift
    local ARGS="$*"

    # shellcheck disable=SC2059:PIDFILE

    [something or other]
}
```

**Comment by FransUrbo** [comment]
… and extending that to multiple variables:

```
# shellcheck disable=SC2059:PIDFILE;VAR1;VAR2;[etc],SC2001:VAR19

**Comment by izabera** [comment]
In your example you want `"$cmd" -p "$pidfile" "$@"` (_quoted!_)
If you want to save it in a local variable, use an array: `local args=("$@")`

**Comment by FransUrbo** [comment]
That is both a bashism and a "Assigning an array to a string"…

```
my_function() {
    local PIDFILE="$1"  ; shift
    local CMD="$1"      ; shift
    local ARGS=("$@")

    "$CMD" -p "$PIDFILE" "$ARGS"
}
```

```
/tmp/tst.sh:6:16: warning: #!/bin/sh was specified, but arrays are not standard. [SC2039]
/tmp/tst.sh:8:27: warning: Expanding an array without an index only gives the first element. [SC2128]
```

And if I use <code>local ARGS="$@"</code> I get:

```
/tmp/tst.sh:6:16: warning: Assigning an array to a string! Assign as array, or use * instead of @ to concatenate. [SC2124]
```

**Comment by izabera** [comment]
You're using `local`, I guess you can call it a bashism since it's not supported in posix sh.
And to expand an array use `"${array[@]}"`

**Comment by FransUrbo** [comment]
Because I'm using bourne shell, It shouldn't tell me about the "Assign as array" (because the're not available).

**Comment by lucy** [comment]
```sh
my_function() {
    PIDFILE="$1" ; shift
    CMD="$1"     ; shift
    "$CMD" -p "$PIDFILE" "$@"
}
```
is posix

**Comment by FransUrbo** [comment]
<code>my_function /bin/ls "${ARGS[@]}"</code>:

```
In /tmp/tst.sh line 11:
my_function /bin/ls "${ARGS[@]}"
                     ^-- SC2039: #!/bin/sh was specified, but array references are not standard.
```

**Comment by FransUrbo** [comment]
Sorry, that was in the wrong place, wrong example. But still. No arrays in bourne shell.

**Comment by FransUrbo** [comment]
@lucy That still is quoted and will be treated as ONE option, not FOUR… It passes the shellcheck tests though. But it's wrong programmatically...

**Comment by izabera** [comment]
Don't merge multiple args in a variable, pass them separate arguments.

**Comment by FransUrbo** [comment]
@izabera That might work in this simple example, but I have a couple of _real_ code where it's required to do it this way. I really don't want one separate variable for each of the options. It's a pain to maintain and extend...

**Comment by lucy** [comment]
```
my_function() {
    local PIDFILE="$1" ; shift
    local CMD="$1"     ; shift
    "$CMD" -p "$PIDFILE" "$@"
}
p() {
	printf '%s\n' "$@"
}
my_function pid p arg1 arg2 arg3
```
prints `-p\npid\narg1\narg2\narg3\n` in dash
"$@" has special expansion that can be used to emulate arrays and get proper splitting

**Comment by FransUrbo** [comment]
@lucy and that's your problem. It should print:

```
# /tmp/tst.sh 
-p
pid
arg1
arg2
arg3
```

As I've said, if you quote <code>$@</code>, it will be treated as ONE command, not multiple...

**Comment by izabera** [comment]
@FransUrbo you could at least test her code, you know

**Comment by FransUrbo** [comment]
No, hang on… Something is amiss here. @izabera Yes, you're right… I need to look at this some more.

**Comment by FransUrbo** [comment]
That exact code (new cut-and-paste!) will give me the output I gave...

**Comment by FransUrbo** [comment]
However, there's a difference here…@lucy is here calling <code>my_function()</code> with each option as separate. I'm calling them as one string.

**Comment by FransUrbo** [comment]
Even if I use <code>/bin/dash</code>, I still get one option per line (if not quoting the args).

**Comment by FransUrbo** [comment]
Also, come to think of it, _I_ think using arguments to a function (without saving it) is bad coding. It will most likely work most of the time, but sooner or later one modifies that (<code>$@</code> and/or <code>$*</code>) and then one uses the wrong values.

I believe in saving the values in the very beginning and then using that. It's … "cleaner" some how.

**Comment by deryni** [comment]
`"$@"` expansion as multiple words is posix. That's the correct way to handle this sort of thing.

If you can't do that then you get to deal with the fact that shellcheck is going to tell you that you should.

I agree with you that a variable specifier on disabling comments is a good idea though.

Stuffing multiple options into a single string is just a bad idea and you really shouldn't do it. It isn't avoidable in some cases if you don't have arrays (well technically you always have one, the positional parameters, but using that can be hard/impossible) but shellcheck is still, to my mind, correct to complain about the usage.

**Comment by FransUrbo** [comment]
I very often create a config file, like something like this:

```
# Extra arguments to "some command"
EXTRA_ARGS="-a -b -c -d"
```

and then source that in the script

```
#!/bin/sh

. /some/dir/config

[automatically/programmatically figure out CALCULATED_ARGS]

"some command" $CALCULATED_ARGS $EXTRA_ARGS
```

I think this is very common, and correct in every way. Quoting these variables is programmatically wrong, which is the point I'm trying to make.

**Comment by deryni** [comment]
It is not "correct in every way". It doesn't work if any arguments need to have embedded spaces or need shell glob metacharacters and it *cannot* be made to support that.

Yes, despite that, it is quite common and works for many cases. That doesn't make it less wrong though.

**Comment by FransUrbo** [comment]
Ok, fair enough.

But the config example is very common and correct here. UNLESS, as you say, you have an argument that contain spaces etc. But that should be used as:

```
# Extra arguments to "some command"
# NOTE: Remember to quote args containing spaces
EXTRA_ARGS="-a -b -c -d 'extra arg'"
```

**Comment by deryni** [comment]
That won't work. Those single quotes make it through to the application and the string is two words/arguments and the single quotes **don't** prevent shell expansion of the contents of the single quoted string.

**Comment by FransUrbo** [comment]
Ok, you're right. That was a bad example probably, but still.

**Comment by FransUrbo** [comment]
```
#!/bin/sh                                                                                                                                                     

my_function() {
    local CMD="$1"      ; shift
    local DIR="$1"      ; shift
    "$CMD" $@
}

ARGS="-C -F -l -a"
my_function /bin/ls "/tmp/some dir" "$ARGS"
```

**Comment by FransUrbo** [comment]
But I think we're getting of topic… I think it's very valid to WANT word splitting on variables in some cases.

But I agree that it might be … "impossible" to programmatically test that with <code>shellcheck</code>, so the ignore "thingie" seems to be a valid functionality.. ?

**Comment by deryni** [comment]
Yes, I think being able to specify a variable for which directive applies is useful functionality personally.

**Comment by FransUrbo** [comment]
I'm just not sure how to specify the <code>env</code> case in the initial comment though. Maybe a line number?

```
# shellcheck disable=SC2046:PIDFILE;VAR1;VAR2;+1;[etc],SC2001:VAR19,SC2034:VAR20,SC2010:+1
```

As I would read this, if put in a function for example, is:

* Ignore <code>SC2059</code> (*Quote this to prevent word splitting*) errors/warnings/notes for variable <code>PIDFILE</code>, <code>VAR1</code>, <code>VAR2</code> and whatever's on the next line.
* Ignore <code> SC2001</code> (*See if you can use …*) for <code>VAR19</code>
* Ignore <code> SC2034</code> (*Unused variable*) for <code>VAR20</code>
* Ignore <code>SC2010</code> (*Don't use ls | grep*) on the following line (I have some old code that isn't worth rewriting of fixed :)

**Comment by FransUrbo** [comment]
Actually, stacking all this in ONE line like this might be a little overzealous. One line per disable is perfectly fine by me.

**Comment by koalaman** [comment]
The [wiki for SC2086](https://github.com/koalaman/shellcheck/wiki/Sc2086) discusses this exact case as an exception to the rule, and suggests using a function instead. 

Directives are scoped to only the immediately following command, so it's generally safe. I.e. if you do

    #shellcheck disable=SC2046
    unset $(env | grep ^FSTAB_ | sed 's,=.*,,')
    touch $(date +%F)

you'll still get a SC2046 warning for the `$(date +%F)`.

I don't entirely see the utility in manually verifying and marking a variable as word splitting safe vs marking the command word splitting safe (by disabling SC2086 on that command). Is the same code called multiple times with just a few arguments changed? Use a function, that's what they're for.

Another point of interest is that since unquoting variables as the command name is usually intentional (very much unlike arguments), ShellCheck doesn't warn for e.g.

    make="make -j 4"
    $make install

**Comment by deryni** [comment]
The use case for variable-specific disabling is that it means you can mark the *one* variable on the complex line that needs to be word-split and not worry about someone later adding, incorrectly, another variable without quotes and not seeing an error about it (or an erroneous edit getting rid of necessary quotes on an existing variable).

On a complex line with multiple expansions that's a useful safety net.

Consider this code:

    #!/bin/sh                                                                                                                                                     

    my_function() {
        local CMD="$1"      ; shift
        local DIR="$1"      ; shift
        "$CMD" "$DIR" $@
    }

    ARGS="-C -F -l -a"
    my_function /bin/ls "/tmp/some dir" "$ARGS"

It would be nice to be able to silence the warning about `$@` without allowing someone to accidentally remove the quotes from `"$DIR"` without getting a warning.

**Comment by koalaman** [comment]
Not quoting arrays is actually a separate warning from not quoting variables, but I see what you mean. 

`bnfc` has a feature where you can specify the number of expected errors, and only get a warning if your changes introduces more. Maybe something like that could be a general way to solve that for all types of warnings?

e.g.

    # shellcheck disable=SC2046:1
    make $(get_flags) "$(get_target)"

so that if you unquote `$(get_target)`, you have 2 SC2046s, and therefore get a warning.

**Comment by deryni** [comment]
A count would cover the same uses but is, I think, strictly less useful here as, I believe, shellcheck knows the variable names and could therefor just match against them and you do really want to limit it to the specific allowed instance and not *any* instance.

Swapping the quotes should be an error as should removing the `$@` (or whatever) usage entirely and replacing with some other unquoted variable.

luacheck does this with its [`--ignore`, `--enable` and  `--only` flags](http://luacheck.readthedocs.org/cli.html#patterns) for example.

**Comment by b-long** [comment]
Based on [this StackOverflow answer](http://stackoverflow.com/a/307735/320399), I wanted to use the following syntax to ensure an environment variable is set: 

```
: ${FOO:?"The environment variable 'FOO' must be set and non-empty"}
```
This seems to work just fine to ensure `$FOO` is set.  However, when I shellcheck it (with version 0.4.1), I get this message :
```
In script.sh line 13:
: ${FOO:?"The environment variable 'FOO' must be set and non-empty"}
  ^-- SC2086: Double quote to prevent globbing and word splitting.
```

Any advice on what I should do in this case?

**Comment by koalaman** [comment]
@b-long if not for style, I'd quote it due to the edge cases mentioned on https://github.com/koalaman/shellcheck/issues/480

**Comment by b-long** [comment]
@koalaman Thank you!  Your comment made me take a second look at this.  I didn't think it was possible to quote the variable (shell syntax still baffles me, thus I use shellcheck).  For future Googlers, this syntax that can be used is : 

```
: "${FOO:?"The environment variable 'FOO' must be set and non-empty"}"
```
Note the use of double-quotes inside and outside of `${ ... }`.

**Comment by deryni** [comment]
@b-long The inner quotes on that expansion are unnecessary. According to [spec](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_02) the contents of _word_ in `${parameter:?[word]}` are only subject to `tilde expansion, parameter expansion, command substitution, and arithmetic expansion` but not word-splitting or globbing.

**Comment by b-long** [comment]
:+1: Thanks @deryni ! Lots to learn with shell programming :)

**Comment by kousu** [comment]
I have a different case:

```
    kill $(jobs -p) 2>/dev/null || true
         ^--------^ SC2046: Quote this to prevent word splitting.
```

Quoting this would be incorrect: jobs -p outputs a list, I want to pass that *as a list* to kill.

**Comment by mminot-yseop** [comment]
Regarding `jobs -p`, I guess you could feed that into `readarray` and then give the array’s items to `kill`.

**Comment by eero-t** [comment]
What about this syntax:
```
# shellcheck disable-variable=SC2086
ARGS="-C -F -l -a"
```

And this being accepted *only* when ARGS content has:
* nothing from external sources (directly or indirectly)
* no glob characters
?

**Comment by mminot-yseop** [comment]
> What about this syntax:
> 
> ```
> # shellcheck disable-variable=SC2086
> ARGS="-C -F -l -a"
> ```
> 
> And this being accepted _only_ when ARGS content has:
> 
>     * nothing from external sources (directly or indirectly)
> 
>     * no glob characters
>       ?

The point (IMHO) is that there are still no advantages in that method with respect to using an array, so you’d just be shooting yourself in the foot with no added value. It’s less maintainable ’cause another contributor could then be less cautious and add weird stuff in there, while with arrays you nearly always have to ask yourself waaaay fewer questions about what would be robust or not.

Also, not sure why you mention globbing characters 🤔 Wouldn’t they be escaped due to the quotes anyway?

**Comment by mminot-yseop** [comment]
> IMHO more reasonably suggestion would have been to use a wrapper function

Yeah I kinda like that. Wouldn’t go as far as to say that it’s always better, but it’s a nice trick.
Note however that `return $?` is nearly always overkill since by default the exit status of functions is the status of the last command it ran. You could remove that line I think.

> verbose / ugly / indirect

Dunno. Functions with good names are cool.

> less cautious contributors would get errors about non-robust changes

Only if shellcheck is run by CI/CD or somehow made mandatory within the company or project. But as far as I know, it’s often just used by some devs’ own initiative on their machine.

Wow those globbing examples are kinda weird indeed xD
Note sure what `ARGS="-a -a" ls $ARGS ?foobar` is supposed to achieve. The assignment will only be visible within the `ls` command, and the arguments in `ARGS` won’t be passed on to `ls`, unlike in `ARGS="-a -a"; ls $ARGS ?foobar` (as two separate instructions):

```bash
  $ set -x

  $ ARGS="-a -a" ls $ARGS ?foobar
+ ARGS='-a -a'
+ ls --color=auto '?foobar'
# No -a -a ↑
ls: cannot access '?foobar': No such file or directory

# ≠

  $ ARGS="-a -a"; ls $ARGS ?foobar
+ ARGS='-a -a'
+ ls --color=auto -a -a '?foobar'
ls: cannot access '?foobar': No such file or directory
```

But maybe I’m misinterpreting your example.

**Comment by eero-t** [comment]
Example was not intended to do anything else besides being short one-liner showing the shellcheck warnings.

Good point about `return $?` being redundant.  For some reason my shellcheck (v0.7.2) did not comment about it?

>> less cautious contributors would get errors about non-robust changes

> Only if shellcheck is run by CI/CD or somehow made mandatory within the company or project.

I see. That's a valid concern.  But so is making asking perfectly fine scripts to be made non-POSIX, because that can blow up in future too.

Maybe shellcheck should always show warning about it needing to be in CI/CD pipeline until one adds a directive to silence that warning?  (only partly joking)

**Comment by mminot-yseop** [comment]
> But so is making asking perfectly fine scripts to be made non-POSIX, because that can blow up in future too.

Wouldn’t it be a huge hassle to make everything POSIX? I kinda do it when I avoid `echo` (once had a Mac user truncating release notes upon generation because of a `\c` haha), or when using `find` (some machines yell if you try to use the implicit `.` path without typing it in), but I really doubt Bash arrays are going anywhere anytime soon. Too many things would blow up all around the world. And they’re soooo versatile and convenient. I don’t think I’d like Bash as much as I do if they weren’t there.

**Comment by mminot-yseop** [comment]
More pragmatically:

* the shellcheck devs probably have lots of stuff to do, and adding exceptions that require to analyze the contents of pseudo-array strings may be annoying, dunno.
* as a dev, I hate seeing unprotected parameter expansion because it forces me to ask myself if this is “normal” or just the result of some dev being too lazy to type a bunch of quotes. It tends to lead to an automatic no-no in pull requests, unless the coder can provide me with a proof that things cannot be done in a reasonably easy way without non-protection (and this has yet to happen).

**Comment by Fernal73** [comment]
I also don't see why exit or return statements with return codes specified by variables need the variables quoted. Isn't it obvious that these are integer variables?

**Comment by mminot-yseop** [comment]
> I also don't see why exit or return statements with return codes specified by variables need the variables quoted. Isn't it obvious that these are integer variables?

`IFS` could contain digits.

```bash
  $ f() { IFS=2; x=123; return $x; }; f
bash: return: too many arguments
```

**Comment by Fernal73** [comment]
> 

Hmm... I seem to have missed that purpose in this rule. It doesn't flag variables declared as integer such as in:
``` bash
declare -i ret=0

return $ret
```

**Comment by SwuduSusuwu** [comment]
[Had to suppress this for simple script](https://github.com/SwuduSusuwu/SubStack/pull/39/commits/e9b22a70d4f193f69bd7674ca828f1d19a33f921). None of the alternatives above will do:
- Can not use `bash`-specific code, as the [project rules are to use portable code (to support Windows, OSX, Solaris, Linux, Unix)](https://github.com/SwuduSusuwu/SubStack/tree/experimental?tab=readme-ov-file#Sh-source).
  - This is a cross platform lib which has as one of its major use cases the fact which it just requires _C++11_ and `/bin/sh`, with specific contracts that embedded tools can use this.
- Can not turn `CTAGS_DEFAULT` into a function, as its whole purpose was to set a single variable diagnostic messages + function calls both use.
- Can not turn the `for` loop into a function, as it must accept a first variable (which can have numerous options) plus numerous paths.
